### PR TITLE
Object level cipher support

### PR DESCRIPTION
**Pull request by srinibas-zededa**
1. controller Cert API, pull task, based on sha mismatch trigger
2. eve Node Ecdh Cert attest API, push task based on sha mismatch trigger
3. cipher event trigger to push, the events to contoller
4. Persistent CipherContextStatus to handle nim/wifi-cred data and associated refactoring
5. tpmmgr to maintain ecdh certificate as persistent data
6. refectoring the zedagent to handle ciper orchestration on config/status template

### CODE REVIEW

File: pkg/pillar/cipher/handlecipher.go [hunk 1]
```diff
@@ -203,10 +157,30 @@ func decryptCipherBlockWithECDH(ctx *DecryptCipherContext,
...
+func lookupEveNodeCert(ctx *DecryptCipherContext,
+	status *types.CipherContextStatus) ([]byte, error) {
+
+	// validate the config
+	if len(status.DeviceCert) == 0 ||
+		status.HasError() {
+		return []byte{}, errors.New("invalid certificate")
+	}
+	// get the private key, for software Ecdh
+	sub := ctx.SubEveNodeCertConfig
```
Comment (eriknordmark) [inline pkg/pillar/cipher/handlecipher.go:171]: But this doesn't work if there is a tpm. Hence the functions which pass around pvtKey are all confusing; we just need to tell the actual decrypt code that it should use the device ECDH key, and get it from the TPM or read it from a file if there is no TPM. Hence no need to add a pvtKey argument AFAICT.

File: pkg/pillar/evetpm/tpm.go [hunk 2]
```diff
@@ -123,6 +123,12 @@ func FileExists(filename string) bool {
 	return err == nil
 }
 
+// remove the file
+func FileRemove(filename string) {
+	os.Remove(filename)
+	return
+}
+
```
Comment (eriknordmark) [inline pkg/pillar/evetpm/tpm.go:131]: This looks like a useless and unused function.

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go [hunk 3]
```diff
@@ -1101,16 +1129,16 @@ func Run(ps *pubsub.PubSub) {
 		ctx.subAttestNonce = subAttestNonce
 		//subAttestNonce.Activate()
 
-		pubAttestCert, err := ps.NewPublication(
+		pubEcdhCertConfig, err := ps.NewPublication(
```
Comment (eriknordmark) [inline pkg/pillar/cmd/tpmmgr/tpmmgr.go:1132]: This is used to publish both the ECDH and attest cert, right? If so it is misnamed.

File: pkg/pillar/cmd/zedagent/handlecertconfig.go [hunk 4]
```diff
@@ -167,51 +313,425 @@ func certsTask(ctx *zedagentContext, triggerCerts <-chan struct{}) {
...
+// Eve node certificate, check whether there is a Sha mismatch
+// to trigger the post request
+func handleEveNodeCertsSha(ctx *zedagentContext,
+	config *zconfig.EdgeDevConfig) {
+	sumHash := hex.EncodeToString(eveNodeCertsHash)
+	certHash := config.GetNodecertConfighash()
+	if sumHash != certHash {
+		triggerEveNodeCertsEvent(ctx)
+	}
+}
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handlecertconfig.go:592]: Why do we need this?
The publish from tpmmgr will only publish if there is a change, hence when handleModify is called in zedagent we know that there was a change.

**Comment by srinibas-zededa** [comment]
I am in the process of generating multiple review pull requests based on
separate functionality.
Regards,
Srinibas

**Comment by eriknordmark** [comment]
> I am in the process of generating multiple review pull requests based on separate functionality.

I mostly want the required pices for the two items. Refactoring can and should be done separately.
If refactoring is required (and the bar is pretty high for this - not just "looking nicer") for the main work it makes sense doing it as a separate PR before that work. The apply the main work.

**Comment by eriknordmark** [comment]
Superceeded by other PR

**Pull request closed without merging**